util/uuid.lua

```diff
@@ -1,9 +1,49 @@
 -- source: https://gist.github.com/jrus/3197011
 local random = math.random
+
+function mail.search_box(playername, box, uuid)
+	local e = mail.get_storage_entry(playername)
+	for _, m in ipairs(e[box]) do
+		if m.id == uuid then
+		return { time = m.time, from = m.from, to = m.to, cc = m.cc, bcc = m.bcc, subject = m.subject, body = m.body } end
+	end
+	return false
+end
+
+function mail.is_uuid_existing(uuid)
+    for _, k in ipairs(mail.storage:get_keys()) do
+        if string.sub(k,1,5) == "mail/" then
+			local p = string.sub(k, 6)
+			local result
+			local boxes = {"inbox", "outbox", "drafts", "trash"}
+			for _, b in ipairs(boxes) do
+				result = mail.search_box(p, b, uuid)
+				if result then return result end
+			end
+		end
+    end
+    return false
+end
+
+function mail.are_message_sames(a, b)
+	return a.time == b.time
+	   and a.from == b.from
+	   and a.to == b.to
+	   and a.cc == b.cc
+	   and a.bcc == b.bcc
+	   and a.subject == b.subject
+	   and a.body == b.body
+end
+
```

Comment:
Pretty sure these belong to migrate.lua, not here. I mean these probably shouldn't be needed or used for anything besides one time repairs and should also be local instead of exposed through global `mail`.

Reply:
`mail.is_uuid_existing(uuid)` should NOT be used in `util/uuid.lua`.
You should NOT have to check newly generated UUID's.

Reply:
I know that the probability is very very ridiculous, but there is still a chance of two generated UUIDs might be the same. It is just here to avoid it.

Reply:
Yeah if either RNG is very bad or if you generate few exabytes of UUID's then you start seeing some percentages you can actually write down readable without scientific notation.

But problem is that if RNG isn't very badly broken it would take hundreds of years non stop generating UUID's and you still don't have any meaningful chance of getting random collision.

If RNG is that badly broken then fix it.

Reply:
> I know that the probability is very very ridiculous, but there is still a chance of two generated UUIDs might be the same. It is just here to avoid it.

As @S-S-X said: this should not happen and you should not have to mitigate against this, the bug (if there even is one) must be elsewhere.

if uuid's really were that bad then every software company in the world would have a major problem, just saying...

Reply:
Done in c052e59

Reply:
> Hmm okay yes, there was never issue but indeed it's not really good, notably on balance benefit/risk. On « big » servers it can take so much time for still many millions of chance under a potential duplicate.

About balance benefit/risk, there's virtually zero risk. I get that it isn't easy to understand mind bogglingly small chances but really here risk translates to zero.

There's a lot higher risk that complicated implementation causes another bug that makes whole thing useless.

To put numbers another way, you'd assume each UUID weights about 1.5 milligrams (two average snowflakes weight about this much, or ~10 sand grains). You then take all possible UUIDs we can generate and put them together, you'd get object that weights about the same as our sun.

If you take some few trillions of UUIDs from that pile it would be impossible to notice that anyone touched it. In terms of mail database, if you had that many mails it would take some 250 terabytes to one petabyte on your server storage and percentage you used up is about 0.000000000000000000000001505%.

Yeah, I know sun is also mind bogglingly large sorry about that. But anything else would make single unit mind bogglingly small for comparisons. It is pretty much impossible to make up anything that is easy to understand.

I hope I got numbers right and bit more accurate than last time, I believe so.